### PR TITLE
feat: add business field mutation to form-level

### LINKED ISSUE #6085: [REVIEW] Business Info (GST Reg No, Address) should be extended to form-level
The `GST Reg No`, and `Business Address` are read during invoice generation on the payment page (after responder has successfully paid and we've received the receipt url from stripe).

This two fields are currently stored and read from form-level details. However, as we expect large agencies to have different `GST Reg No`, and `Business Address`, we want to add support to store and alter them at the form-level.

Storing at the form-level also supports the edge case where the form is **transferred to a different agency**, but the payment channel information should still remain the same.

The agency-level information will then be used as a prefill when the admin first.

Context:
- #6078

### PR DESCRIPTION
**Pull request by KenLSM**
## Problem

The GST Reg No, and Business Address are read during invoice generation on the payment page (after responder has successfully paid and we've received the receipt url from stripe).

This two fields are currently stored and read from form-level details. However, as we expect large agencies (i.e., MOH, MOE) to have different GST Reg No, and Business Address, we want to add support to store and alter them at the form-level.

Storing at the form-level also supports the edge case where the form is transferred to a different agency, but the payment channel information should still remain the same.

Closes #6085

## Solution

Introducing form-level overrides.

**Breaking Changes** 

- [x] No - this PR is backwards compatible  

**Features**:

- Adds options for form admins to enter their own business information at the form level
- Invoice will respect form-level overrides before reading agency-level information

## Before & After Screenshots

**BEFORE**:

Screenshot 2023-05-03 at 12 37 17 AM

**AFTER**:

Screenshot 2023-05-03 at 12 37 30 AM

Screenshot 2023-05-03 at 12 37 04 AM

Screenshot 2023-05-03 at 10 56 42 AM

## Tests

**Form Level Overrides**
Payment form business fields can be changed
- [ ] change `gstRegNo` field to `foobar`, expect that a toast appears confirming change
- [ ] change `gstRegNo` field to ` foobar` (with leading/trailing spaces), expect that a toast appears confirming change
- [ ] field is populated with `foobar` after a refresh

Agency default is shown as placeholder if form business info is not present
- [ ] change `gstRegNo` field to `` (empty string), expect that a toast appears confirming change and showing placeholder message
  - [ ] field is populated with agency defaults (OGP's agency gstRegNo/address is used)
  - [ ] data retrieved for `form.business.gstRegNo` and `*.address` is ``

Generated Invoice should use form-level business field if present
- [ ] prepare payment form with form-level overrides on `gstRegNo` + `address` fields with `foobar`
- [ ] generated invoice should show `GST Reg No: foobar` and `Address: foobar`

Business Info Fields will trim
- [ ] change `gstRegNo` and `address` field to ` foobar` (with leading/trailing spaces), expect that a toast appears confirming change
- [ ] field is populated with `foobar` (without leading/trailing spaces) after a refresh

**Regression tests**
- [ ] payment forms without overrides
  - [ ] other fields (Form Name) should be able editable
  - [ ] generated invoice should fallback to agency's business info

### CODE REVIEW

File: frontend/src/features/admin-form/settings/components/PaymentSettingsSection/BusinessInfoSection.tsx [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  initialValue,
+  handleMutation,
+}: BusinessFieldInputProps): JSX.Element => {
+  const [value, setValue] = useState(initialValue)
+
+  const inputRef = useRef<HTMLInputElement>(null)
+
+  const handleValueChange: ChangeEventHandler<HTMLInputElement> = useCallback(
+    (e) => {
+      setValue(e.target.value.trim())
```
Comment (foochifa) [inline frontend/src/features/admin-form/settings/components/PaymentSettingsSection/BusinessInfoSection.tsx]: Was wondering if we could delay the trimming to before inputting into handle mutation

The reason for this is when we are filling in the address it becomes really unintuitive.

e.g. We will type '36 Robinson Road', however with the current trim at setState it will be '36RobinsonRoad' and we have to go back to the spaces to add them.

File: frontend/src/features/admin-form/settings/components/PaymentSettingsSection/BusinessInfoSection.tsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
+import {
+  ChangeEventHandler,
+  Dispatch,
+  FocusEventHandler,
+  KeyboardEventHandler,
+  SetStateAction,
```
Comment (foochifa) [inline frontend/src/features/admin-form/settings/components/PaymentSettingsSection/BusinessInfoSection.tsx]: Couple of unused imports I think, `Dispatch`, `FocusEventHandler`, `SetStateAction`

File: frontend/src/features/admin-form/settings/components/PaymentSettingsSection/BusinessInfoSection.tsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+          initialValue={settings.business?.gstRegNo || ''}
+          handleMutation={handleGstRegNoMutation}
+        />
+      </FormControl>
+      <FormControl mb="2.5rem" isReadOnly={mutateFormBusiness.isLoading}>
+        <FormLabel description="Leave empty to use your agency defaults.">
+          Business Address
+        </FormLabel>
+        <BusinessFieldInput
+          initialValue={settings.business?.address || ''}
```
Comment (foochifa) [inline frontend/src/features/admin-form/settings/components/PaymentSettingsSection/BusinessInfoSection.tsx:113]: Maybe not for this PR, but we should technically add the default agency values as the initial value right?

**Review by foochifa — COMMENTED** [review]
Just some comments on `BusinessInfoSection`!

File: frontend/src/features/admin-form/settings/components/PaymentSettingsSection/BusinessInfoSection.tsx (hunk 1 above)
Comment (KenLSM) [reply frontend/src/features/admin-form/settings/components/PaymentSettingsSection/BusinessInfoSection.tsx]: Right 👍 , definitely should delay this.

File: frontend/src/features/admin-form/settings/components/PaymentSettingsSection/BusinessInfoSection.tsx (hunk 3 above)
Comment (KenLSM) [reply frontend/src/features/admin-form/settings/components/PaymentSettingsSection/BusinessInfoSection.tsx:113]: Yeah, in another feature ticket. This is ticket is mainly to keep the scope as small as possible to ship this for form admins to change. Prefilling takes a bit more time to think about.

File: frontend/src/features/admin-form/settings/components/PaymentSettingsSection/BusinessInfoSection.tsx (hunk 2 above)
Comment (KenLSM) [reply frontend/src/features/admin-form/settings/components/PaymentSettingsSection/BusinessInfoSection.tsx]: bcde461d

File: frontend/src/features/admin-form/settings/components/PaymentSettingsSection/BusinessInfoSection.tsx (hunk 1 above)
Comment (KenLSM) [reply frontend/src/features/admin-form/settings/components/PaymentSettingsSection/BusinessInfoSection.tsx]: 160ea4cb

Delayed to after user has clicked away from the input, trim to apply just before the value was sent as a patch to BE

File: src/app/modules/payments/stripe.controller.ts [hunk 4]
```diff
@@ -361,28 +361,36 @@ export const downloadPaymentInvoice: ControllerHandler<{
           // convert to pdf and return
           .then((receiptUrlResponse) => {
```
Comment (foochifa) [inline src/app/modules/payments/stripe.controller.ts:362]: Maybe for another ticket/PR,

but was wondering if we might want to abstract out the entire section for converting the receipt to url (line 363-400) conversion to `payment.service.ts`

this may help us to write unit tests to ensure the behaviour is correct e.g.
- will use the form address/gst before agency if available
- will still generate invoice without address/gst (which may be in the future this might be an error?)

It may also be a bit cleaner on our controller 
`.then((receiptUrlResponse) => convertToInvoicePdf(receiptUrlResponse, populatedForm))`

P.s. yikes I pressed single comment instead of review 😅

**Review by foochifa — APPROVED** [review]
LGTM! Thanks Ken, have tested the changes on the trimming too

Some potential additional tests:
- [ ] add trailing whitespace to address, upon blur it should be removed

Also the test case to ensure the agency level defaults are used when form business is empty can be a bit more explicit 
- [ ] enter '' for gst reg no and address, ensure the placeholder 'Using Agency Defaults' is displayed
- [ ] make a payment and generate the invoice, the address and gst reg no should be the agency defaults

And another one for when no user provided address/gst and agency defaults are available
- [ ] for a form with no form level address/gst, and no agency defaults. Ensure an invoice can be generated with no errors

File: src/app/modules/payments/stripe.controller.ts (hunk 4 above)
Comment (KenLSM) [reply src/app/modules/payments/stripe.controller.ts:362]: Yeah makes sense!
